Versioned block support

olegshmuelov opened the pull request:
- Versioned block support for regular/blinded beacon block
- Deprecate `Bellatrix` ===> Introduce `Capella`
- Update tests with new roots
closes #140 

THIS PR split to 2 different PRs:
 - basic support for versioned block -  https://github.com/bloxapp/ssv-spec/pull/208
 - TBD. deprecates older forks & introduce Capella with state comparison.

olegshmuelov commented:
Can be merged only after https://github.com/attestantio/go-eth2-client/issues/50 resolved

alonmuroch commented on go.mod:
```diff
@@ -70,3 +70,5 @@ require (
 	lukechampine.com/blake3 v1.1.7 // indirect
 )
+
+replace github.com/attestantio/go-eth2-client => github.com/olegshmuelov/go-eth2-client v0.0.0-20230321143339-08ced80c995f
```
needs to be replaced to be merged

alonmuroch commented on types/ssz.go:
```diff
@@ -81,6 +83,34 @@ func (b SSZTransactions) HashTreeRoot() ([32]byte, error) {
 	return ssz.HashWithDefaultHasher(b)
 }
 
+// SSZWithdrawals --
```
spec tests

alonmuroch commented on types/testingutils/ssv_msgs.go:
```diff
@@ -307,7 +308,7 @@ var postConsensusBeaconBlockMsg = func(
 	copy(blsSig[:], sig)
 
-	signed := bellatrix.SignedBeaconBlock{
+	signed := capella.SignedBeaconBlock{
```
add the new cappella fields like withdrawals

olegshmuelov commented on types/spectest/tests/consensusdata/invalid_bellatrix_blinded_block_validation.go:
```diff
@@ -1,6 +0,0 @@
-package consensusdata
-
-// InvalidBellatrixBlindedBlockValidation tests an invalid consensus data with bellatrix blinded block
```
added spectests under `types/consensusdata`

GalRogozinski commented on types/testingutils/beacon_node.go:
```diff
@@ -486,56 +503,84 @@ func (bn *TestingBeaconNode) GetBeaconNetwork() types.BeaconNetwork {
+	var r [32]byte
+	switch block.Version {
+	case spec.DataVersionCapella:
+		if block.Capella == nil {
```
should we also check if the sig is nil?
not sure

olegshmuelov replied:
not needed will fail validation

GalRogozinski commented on types/testingutils/beacon_node.go:
```diff
@@ -486,56 +503,84 @@ func (bn *TestingBeaconNode) GetBeaconNetwork() types.BeaconNetwork {
+
+	switch block.Version {
+	case spec.DataVersionCapella:
+		if block.Capella == nil {
```
ditto sig nil check?

olegshmuelov replied:
ditto

GalRogozinski commented on types/generate.go:
```diff
@@ -10,10 +10,10 @@ package types
-//go:generate go run .../fastssz/sszgen --path beacon_types.go --include $GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.1/spec/phase0 --exclude-objs BeaconNetwork,BeaconRole
+//go:generate go run .../fastssz/sszgen --path beacon_types.go --include $GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.8-0.20230322092059-6b2aee891155/spec/phase0 --exclude-objs BeaconNetwork,BeaconRole
 
 //go:generate rm -f ./partial_sig_message_encoding.go
-//go:generate go run .../fastssz/sszgen --path partial_sig_message.go --include $GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.1/spec/phase0,./signer.go,./operator.go --exclude-objs PartialSigMsgType
+//go:generate go run .../fastssz/sszgen --path partial_sig_message.go --include $GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.8-0.20230322092059-6b2aee891155/spec/phase0,./signer.go,./operator.go --exclude-objs PartialSigMsgType
 
 //go:generate rm -f ./consensus_data_encoding.go
-//go:generate go run .../fastssz/sszgen --path consensus_data.go --include ./operator.go,./signer.go,./partial_sig_message.go,./beacon_types.go,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.1/spec/phase0,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.1/spec,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.1/spec/altair --exclude-objs Contributions,BeaconNetwo ...
+//go:generate go run .../fastssz/sszgen --path consensus_data.go --include ./operator.go,./signer.go,./partial_sig_message.go,./beacon_types.go,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.8-0.20230322092059-6b2aee891155/spec/phase0,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@v0.15.8-0.20230322092059-6b2aee891155/spec,$GOPATH/pkg/mod/github.com/attestantio/go-eth2-client@ ...
```
:+1:

GalRogozinski commented on types/ssz.go:
```diff
@@ -81,6 +83,34 @@ func (b SSZTransactions) HashTreeRoot() ([32]byte, error) {
+// SSZWithdrawals --
+type SSZWithdrawals []*capella.Withdrawal
+
+func (b SSZWithdrawals) GetTree() (*ssz.Node, error) {
```
is there a reason why we don't use a pointer here?

olegshmuelov replied:
lets derive a task for this

GalRogozinski commented on types/ssz.go:
```diff
@@ -81,6 +83,34 @@ func (b SSZTransactions) HashTreeRoot() ([32]byte, error) {
+	return ssz.ProofTree(b)
+}
+
+func (b SSZWithdrawals) HashTreeRootWith(hh ssz.HashWalker) error {
```
ditto

olegshmuelov replied:
ditto

GalRogozinski commented on types/ssz.go:
```diff
@@ -81,6 +83,34 @@ func (b SSZTransactions) HashTreeRoot() ([32]byte, error) {
+}
+
+// HashTreeRoot --
+func (b SSZWithdrawals) HashTreeRoot() ([32]byte, error) {
```
ditto

olegshmuelov commented on types/spectest/tests/consensusdata/proposer/versioned_block_validation.go:
```diff
@@ -0,0 +1,24 @@
+
+// VersionedBlockValidation tests a valid consensus data with capella block
+func VersionedBlockValidation() *ProposerSpecTest {
+	expectedCdRoot, err := testingutils.TestProposerConsensusData.HashTreeRoot()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	expectedBlkRoot, err := testingutils.TestingBeaconBlock.HashTreeRoot()
+	if err != nil {
```
ditto

olegshmuelov commented on types/spectest/tests/ssz/test.go:
```diff
@@ -0,0 +1,41 @@
+
+	var withdrawals types.SSZWithdrawals
+	switch vBlk.Version {
+	case spec.DataVersionCapella:
+		withdrawals = vBlk.Capella.Body.ExecutionPayload.Withdrawals
+	}
+
+	root, err := withdrawals.HashTreeRoot()
+	require.NoError(t, err)
```
done

olegshmuelov commented on types/spectest/tests/ssz/test.go:
```diff
@@ -0,0 +1,41 @@
+
+	var withdrawals types.SSZWithdrawals
+	switch vBlk.Version {
+	case spec.DataVersionCapella:
+		withdrawals = vBlk.Capella.Body.ExecutionPayload.Withdrawals
+	}
+
+	root, err := withdrawals.HashTreeRoot()
+	require.NoError(t, err)
```
duplicate

olegshmuelov commented on types/spectest/tests/consensusdata/proposer/versioned_blinded_block_unknown_version.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/bloxapp/ssv-spec/types/testingutils"
+)
+
+// VersionedBlindedBlockUnknownVersion tests a valid consensus data with unknown block
```
done

olegshmuelov commented on types/spectest/tests/ssz/withdrawals_marshaling_validation.go:
```diff
@@ -0,0 +1,23 @@
+
+	return &SSZSpecTest{
+		Name:         "ssz withdrawals marshalling",
+		Data:         testingutils.TestProposerConsensusDataByts,
```
This way we can see the whole picture

olegshmuelov commented on types/spectest/tests/consensusdata/proposer/versioned_block_spectest.go:
```diff
@@ -0,0 +1,110 @@
+			blkSSZ, err = vBlk.Capella.MarshalSSZ()
+			require.NoError(t, err)
+		}
+		require.EqualValues(t, test.DataBlk, blkSSZ)
```
done

GalRogozinski commented on types/spectest/tests/consensusdata/proposer/versioned_blinded_block_validation.go:
```diff
@@ -0,0 +1,25 @@
+
+// VersionedBlindedBlockValidation tests a valid consensus data with capella blinded block
+func VersionedBlindedBlockValidation() *ProposerSpecTest {
+	expectedCdRoot, err := testingutils.TestProposerBlindedBlockConsensusData.HashTreeRoot()
```
I think what we can do is have SpecTests that test the `HashTreeRoot()`for different objects.
Those tests will have the expected results hardcoded.

Then we can assume that no need to hardcode expected roots for the rest of the tests.

GalRogozinski replied:
https://bloxxx.atlassian.net/browse/IO1-3182

y0sher commented on types/consensus_data.go:
```diff
@@ -156,36 +147,30 @@ func (ci *ConsensusData) GetAttestationData() (*phase0.AttestationData, error) {
-	if err := ret.UnmarshalSSZ(ci.DataSSZ); err != nil {
-		return nil, errors.Wrap(err, "could not unmarshal ssz")
+func (ci *ConsensusData) GetBlockData() (*spec.VersionedBeaconBlock, ssz.HashRoot, error) {
+	switch ci.Version {
```
if we're basically deprecating Bellatrix, this is a preparation to support future version ?

olegshmuelov replied:
exactly, so if for example next fork [`Deneb`](https://crypto.news/days-after-shanghai-ethereum-developers-readying-for-key-eip-and-deneb-upgrade/) will introduce new block, we just need to upgrade the eth2-go-client package and adjust the code to use the `Deneb` beacon block

y0sher commented on types/testingutils/beacon_node.go:
```diff
@@ -2,122 +2,137 @@ package testingutils
+	return res
+}
+
+var Transactions = func() []bellatrix.Transaction {
```
don't we have `capella.Transaction`? or this is a mistake?

olegshmuelov replied:
no, the transactions introduced in Bellatrix fork

alonmuroch commented on ssv/proposer.go:
```diff
@@ -190,29 +193,21 @@ func (r *ProposerRunner) ProcessPostConsensus(signedMsg *types.SignedPartialSign
 		copy(specSig[:], sig)
 
 		if r.decidedBlindedBlock() {
-			data, err := r.GetState().DecidedValue.GetBellatrixBlindedBlockData()
```
not sure how significant it is but versioned block is "wasteful" since ssz will fill in zero data in some fields

olegshmuelov replied:
If I understand your comment correctly, it doesn't make sense because we are only marshaling SSZ for specific blocks like Bellatrix, Capella, etc., rather than the entire VersionedBlock

alonmuroch commented:
@olegshmuelov @y0sher @GalRogozinski @lior-blox all the test root changes, anyone went one by one to see they actually are correct? A lot was changed in this PR so it needs to be checked one-by-one.


@olegshmuelov do we have a test there actually testing different versions of the block? seems we just changed everything to cappella but no test for actually multi version support

y0sher commented:
@alonmuroch  wouldn't the tests not pass if they aren't correct? I believe @olegshmuelov went one by one and change them. one of us can double check if that is needed and it can be wrong.

@olegshmuelov about the second point I agree we should have a test that covers the case where we got a block with a different version than Cappella.

alonmuroch commented:
> @alonmuroch wouldn't the tests not pass if they aren't correct? I believe @olegshmuelov went one by one and change them. one of us can double check if that is needed and it can be wrong.
The tests pass but we don't know against which post root as we changed them. If the post root is not correct it will pass anyhow. That's why we introduced state comparison but it's not yet implemented here.
I suggest someone to go root by root with breakpoints to see it actually produces the correct state

GalRogozinski commented:
Let's go over one by one again when we add state comparisons @alonmuroch

GalRogozinski commented:
@olegshmuelov 
After discussing with Alon and @lior-blox, it seems that the best way forward is to split the PR this way:
1. Add versioned block support but don't upgrade to Capella.
2. Upgrade to Capella with state compariosn feature. This way we ensure the state rrots are correct and we add more state comparison tests.

olegshmuelov commented on go.mod:
```diff
@@ -3,7 +3,7 @@ module github.com/bloxapp/ssv-spec
 
 require (
-	github.com/attestantio/go-eth2-client v0.15.1
+	github.com/attestantio/go-eth2-client v0.15.8-0.20230322092059-6b2aee891155
```
The PR was ready before the `go-eth2-client` tag was released
https://github.com/bloxapp/ssv-spec/pull/171#issuecomment-1478337320

olegshmuelov commented:
> @olegshmuelov @y0sher @GalRogozinski @lior-blox all the test root changes, anyone went one by one to see they actually are correct? A lot was changed in this PR so it needs to be checked one-by-one.
> 
> @olegshmuelov do we have a test there actually testing different versions of the block? seems we just changed everything to cappella but no test for actually multi version support

1. Yes it was checked by me, also we will probably add state comparison tests to this PR as well before merging it
2. First this PR was introduced with all block versions for regular/blinded blocks but then was requested by you to remove it. 
This why we deprecated all Bellatrix block tests and switched to use Capella
relevant [commit](https://github.com/bloxapp/ssv-spec/pull/171/commits/d021b1e5782bc2d8d48122588dcf80c246be92df#diff-c02f04705d9ade24a834f46564e331fbeef4cf8fa826c4a3071b0f946f177fe1):

GalRogozinski commented:
Will be split to several different PRs
